src/render/draw_custom.js

```diff
@@ -0,0 +1,58 @@
+// @flow
+
+export default drawCustom;
+
+import DepthMode from '../gl/depth_mode';
+import StencilMode from '../gl/stencil_mode';
+import {prepareOffscreenFramebuffer, drawOffscreenTexture} from './offscreen';
+
+import type Painter from './painter';
+import type SourceCache from '../source/source_cache';
+import type CustomStyleLayer from '../style/style_layer/custom_style_layer';
+
+function drawCustom(painter: Painter, sourceCache: SourceCache, layer: CustomStyleLayer) {
+
+    const context = painter.context;
+    const implementation = layer.implementation;
+
+    if (painter.renderPass === 'offscreen') {
+
+        const prerender = implementation.prerender;
+        if (prerender) {
+            painter.setCustomLayerDefaults();
+
+            prerender.call(implementation, context.gl, painter.transform.customLayerMatrix());
+
+            context.setDirty();
+            painter.setBaseState();
+        }
+
+        if (implementation.renderingMode === '3d') {
+            painter.setCustomLayerDefaults();
+
+            prepareOffscreenFramebuffer(painter, layer);
+            implementation.render(context.gl, painter.transform.customLayerMatrix());
+
+            context.setDirty();
+            painter.setBaseState();
+        }
+
+    } else if (painter.renderPass === 'translucent') {
+
+        if (implementation.renderingMode === '3d') {
+            drawOffscreenTexture(painter, layer, 1);
```

Comment:
Let's add a way to control this — e.g. instead of `1`, `implementation.opacity === undefined ? 1 : implementation.opacity`.

Reply:
I'd rather not expose this right now.

Opacity for our 3d layers is problematic because we can't get ordering correct for semi-transparent layers. I think we may have to change how this is done and I'd rather not expose something we would have to break.